src/Algebra/Graph/Algorithm/Internal.hs

```diff
@@ -1,3 +1,30 @@
 module Algebra.Graph.Algorithm.Internal where
 
+import           Algebra.Graph
+import qualified Data.DisjointSet as DS
+import qualified Data.Set         as S
 
+-- | Extract disjoint set of connectivity components from the graph
+components :: Ord a => Graph a -> DS.DisjointSet a
+components Empty         = DS.empty
+components (Vertex x)    = DS.singleton x
+components (Overlay x y) = mergeComponents (components x) (components y)
+components (Connect x y) = DS.fromLists [vertexList x ++ vertexList y]
+
+-- | Merge two disjoint sets
+mergeComponents ::
+     Ord a => DS.DisjointSet a -> DS.DisjointSet a -> DS.DisjointSet a
+mergeComponents x y = foldr insertComponent x (DS.toSets y)
+
+-- | Insert a set into the other disjoint set structure:
+-- Firstly, we find all intersections between the given set `x` and any of
+-- disjoint ones. Then we unite all the intersecting sets into one big union
+-- and finally append this big union to all other sets, with which we were not
+-- intersected.
+insertComponent :: Ord a => S.Set a -> DS.DisjointSet a -> DS.DisjointSet a
+insertComponent x dsu =
+  DS.fromLists (S.toList (S.unions (x : connected)) : disconnected)
```

Comment:
Converting to lists and back is not a good way to go, but is fine for the prototype. We need to look for a way to efficiently merge `DisjointSet`s.